plugin/semisync/semisync_master.cc

```diff
@@ -513,13 +522,29 @@ void ReplSemiSyncMaster::lock() { mysql_mutex_lock(&LOCK_binlog_); }
 
 void ReplSemiSyncMaster::unlock() { mysql_mutex_unlock(&LOCK_binlog_); }
 
-void ReplSemiSyncMaster::add_slave() {
+void ReplSemiSyncMaster::add_slave(int server_id) {
```

Comment:
The server_id comes from the slave, but uniqueness is to be enforced by an external agent. What happens if this is not true and two slaves have matching server_id, but different UUIDs? It seems depending on this value to separate out slaves could be risky, especially since it could lead to whitelisting of slaves whose UUID is not in the list, but whose server_id matched of a slave whose UUID is in the list.

One could make the argument about UUIDs conflicting too and assume the same system that assigned server_id is the same as that for UUID, so either both server_id+UUID conflict or neither conflict. But there is probably some corner case where server_id conflict, but UUID does not and vice versa.